extends Node

signal quest_added(id: String)
signal quest_started(id: String)
signal quest_completed(id: String)
signal quests_loaded()

func load_day_quests(resource_path: String):
	var script = load(resource_path)
	if script == null:
		push_error("âŒ day quests íŒŒì¼ ë¡œë“œ ì‹¤íŒ¨: " + resource_path)
		return

	var instance = script.new()

	if not instance.has_variable("quests"):
		push_error("âŒ day quests íŒŒì¼ì— quests ë³€ìˆ˜ê°€ ì—†ìŒ")
		return

	var quest_array = instance.quests

	for q in quest_array:
		QuestManager.add_quest(q["id"], q["name"], q["desc"])

	print("ðŸ“Œ Day quests ë¡œë“œ ì™„ë£Œ:", quest_array.size())

var quests: Dictionary = {}



func _ready() -> void:
	load_quests()


# ----------------------
# í€˜ìŠ¤íŠ¸ ì¶”ê°€
# ----------------------
func add_quest(id: String, name: String, desc: String = "") -> void:
	if quests.has(id):
		return

	quests[id] = {
		"name": name,
		"desc": desc,
		"status": "not_started"
	}

	emit_signal("quest_added", id)



# ----------------------
# í€˜ìŠ¤íŠ¸ ì‹œìž‘
# ----------------------
func start_quest(id: String) -> void:
	if not quests.has(id):
		return

	var q: Dictionary = quests[id]

	if q["status"] == "not_started":

		quests[id] = q
		emit_signal("quest_started", id)



# ----------------------
# í€˜ìŠ¤íŠ¸ ì™„ë£Œ
# ----------------------
func complete_quest(id: String) -> void:
	print("ì‹ í˜¸ë°›ìŒ")
	if not quests.has(id):
		return

	var q: Dictionary = quests[id]

	if q["status"] != "completed":
		print("completedìœ¼ë¡œ ë°”ë€œ")
		q["status"] = "completed"
		quests[id] = q
		emit_signal("quest_completed", id)



# ----------------------
# ì¡°íšŒ í•¨ìˆ˜ë“¤
# ----------------------
func is_completed(id: String) -> bool:
	return quests.has(id) and quests[id]["status"] == "completed"


func get_all_quests() -> Dictionary:
	return quests.duplicate(true)


# ----------------------
# ì €ìž¥
# ----------------------
func all_quests_completed() -> bool:
	for id in quests.keys():
		if quests[id]["status"] != "completed":
			return false
	return true


# ----------------------
# ë¡œë“œ
# ----------------------
func load_quests(path: String = "") -> void:
	var p: String = path 

	if not FileAccess.file_exists(p):
		return

	var file := FileAccess.open(p, FileAccess.READ)
	if file == null:
		return

	var text: String = file.get_as_text()
	file.close()

	var parsed = JSON.parse_string(text)

	if typeof(parsed) != TYPE_DICTIONARY:
		return

	if parsed.has("quests"):
		quests = parsed["quests"]
	else:
		quests = {}

	emit_signal("quests_loaded")
